### DISCUSSION TITLE
Help with beacon integration for embrace

### DISCUSSION THREAD

**Discussion by Haxxflaxx** [meow-edit/meow#667] [Q&A]
I'm trying to get embrace to apply to all selections in beacon mode, and after some digging around in how the insert functions handle this I came up with the following:

```emacs-lisp
  (defun da/meow-beacon-embrace ()
    (interactive)
    (meow-beacon-mode -1)
    (setq-local meow--beacon-insert-enter-key last-input-event)

    (call-interactively #'kmacro-start-macro)
    (embrace-commander)
    (when defining-kbd-macro
      (end-kbd-macro)
      (meow--beacon-apply-kmacros-from-insert))

    (meow--switch-state 'beacon))
```

It feels like something like this should work, but for some reason I don't understand, it doesn't.
It fails to record the key presses done within `embrace-commander` and so it just goes to the next selection, runs embrace and stops waiting for input.

If I do the macro recording manually it all works as I expect and applies to all other selections without extra input, so there is something different going on when I'm trying to do this programmatically.

Is there something I'm missing in how macro recording works?

**Comment by DogLooksGood** [comment]
Please pull the latest commit and turn on `meow-keypad-execute-on-beacons`. Then use keypad directly in beacon state.

Well, `embrace-commander` won't work here. Is there some commands to bind directly instead of using the menu?

I guess your code doesn't work because after starting kmacro recording, you don't have the key input to trigger the menu again.

**Reply by Haxxflaxx** [reply]
Ok `meow-keypad-execute-on-beacons` is really cool and makes a lot of things much nicer (suddenly easy `upcase-dwim`).

The problem I run into is that `embrace-commander` takes input for how it should act. An example is (i have it bound to 'l' currently): "l a )" would add parens around the selection. When no selection is active you will also get a prompt for what to act on: "l a w )" would add parens around the word at pt.

This requires the keys to be recorded while interacting with the command so that it can be replayed on the beacons. This is why I thought how insert is handled would be a good thing to model it on.

**Reply by DogLooksGood** [reply]
I tried embrace, and guess there are two cases in general, not just about embrace.

1. a single command takes input in its process, if we wrap it with macro recording, it should work then.
2. a command brings up a keymap, in this case, we don't know where to stop recording.

**Reply by Haxxflaxx** [reply]
I'm back too looking at this again and it seems to be the `read-char` that is the problem in this case.
I've built a minimal failing example:
```emacs-lisp
(defun interactive-function ()
   (read-char "Input: "))

(defun macro-test ()
  (interactive)
  (start-kbd-macro nil)
  (interactive-function)
  (end-kbd-macro))
```

When running `macro-test` and inspecting `last-kbd-macro` then the value is "", like in the case with embrace.
If `read-char` is replaces with `read-string`, then the keys are recorded.
If `end-kbd-macro`is moved out of the function and called after `macro-test` has completed, then the keys are recorded.

I find this very strange. Seems like a timing issue, which it shouldn't be able to be.
